Fix introduce retries if usdc tx fetching exceeds limits

ii-cruz opened the pull request:
Introduces a mechanism to retry the request with a smaller block interval if there is an error.

This is an example of the error message we get where no more tx are fetched afterwards.
`Query timeout exceeded. Consider reducing your block range. Based on your parameters and the response size limit, this block range should work: [0x42b950f, 0x435784f]`

Alchemy restricts log responses to 10k if the block range is >2000. So the safe query filter is tried with the provided block range, and if it fails, it attempts with a 2000 block range. If that still fails, it halves the range. https://www.alchemy.com/docs/deep-dive-into-eth_getlogs

ii-cruz committed:
Introducing safe querying

ii-cruz committed:
Debugging

ii-cruz committed:
Debugging 2

ii-cruz committed:
Error handling

ii-cruz committed:
Removing error parsing. Adopting 2000 range instead

ii-cruz committed:
Fixing logs

ii-cruz committed:
Commenting

ii-cruz committed:
Fixes

sisou commented on src/ethers.ts:
```diff
@@ -914,8 +971,8 @@ export async function launchPolygon() {
-                    client.usdcToken.queryFilter(filterIncoming, startHeight, endHeight),
-                    client.usdcToken.queryFilter(filterOutgoing, startHeight, endHeight),
+                    safeQueryFilter(client.usdcBridgedToken, filterIncoming, startHeight, endHeight),
+                    safeQueryFilter(client.usdcBridgedToken, filterOutgoing, startHeight, endHeight),
```
Copy-past error probably.
```suggestion
                    safeQueryFilter(client.usdcToken, filterIncoming, startHeight, endHeight),
                    safeQueryFilter(client.usdcToken, filterOutgoing, startHeight, endHeight),
```

sisou commented on src/ethers.ts:
```diff
@@ -1175,8 +1232,8 @@ export async function launchPolygon() {
-                    client.usdtBridgedToken.queryFilter(filterIncoming, startHeight, endHeight),
-                    client.usdtBridgedToken.queryFilter(filterOutgoing, startHeight, endHeight),
+                    safeQueryFilter(client.usdcBridgedToken, filterIncoming, startHeight, endHeight),
+                    safeQueryFilter(client.usdcBridgedToken, filterOutgoing, startHeight, endHeight),
```
Same here:
```suggestion
                    safeQueryFilter(client.usdtBridgedToken, filterIncoming, startHeight, endHeight),
                    safeQueryFilter(client.usdtBridgedToken, filterOutgoing, startHeight, endHeight),
```

sisou commented on src/ethers.ts:
```diff
@@ -277,6 +277,63 @@ async function updateUsdtBridgedBalances(addresses: string[] = [...usdtBridgedBa
+    // Subsequent tries halve the range until success or minimum range reached.
+    while (currentStart <= toBlock) {
+        // Attempt to fetch logs in the current range.
+        // Reduce the window size until it no longer crosses breaches alchemy's limits.
```
```suggestion
        // Reduce the window size until it no longer exceeds Alchemy's limits.
```

sisou commented on src/ethers.ts:
```diff
@@ -277,6 +277,63 @@ async function updateUsdtBridgedBalances(addresses: string[] = [...usdtBridgedBa
+        // Reduce the window size until it no longer crosses breaches alchemy's limits.
+        while (true) {
+            try {
+                console.warn('Trying to fetch ', currentStart, currentEnd);
```
```suggestion
                console.warn('Trying to fetch from Alchemy:', currentStart, currentEnd);
```

sisou commented on src/ethers.ts:
```diff
@@ -277,6 +277,63 @@ async function updateUsdtBridgedBalances(addresses: string[] = [...usdtBridgedBa
+                allEvents.push(...eventsChunk);
+                break;
+            } catch (err: any) {
+                const currentRage = currentEnd - currentStart;
```
```suggestion
                const currentRange = currentEnd - currentStart;
```

sisou commented on src/ethers.ts:
```diff
@@ -277,6 +277,63 @@ async function updateUsdtBridgedBalances(addresses: string[] = [...usdtBridgedBa
+                // Try the recommended max range first, otherwise halves the range.
+                if (currentRage > MAX_RECOMMENDED_RANGE) {
+                    currentEnd = currentStart + MAX_RECOMMENDED_RANGE;
+                    console.warn('Query filter failed retrying with rage of 2000.', err, currentStart, currentEnd);
```
```suggestion
                    console.warn('Query filter failed. Retrying with range of 2000.', err, currentStart, currentEnd);
```

sisou commented on src/ethers.ts:
```diff
@@ -277,6 +277,63 @@ async function updateUsdtBridgedBalances(addresses: string[] = [...usdtBridgedBa
+                    currentEnd = currentStart + MAX_RECOMMENDED_RANGE;
+                    console.warn('Query filter failed retrying with rage of 2000.', err, currentStart, currentEnd);
+                } else {
+                    currentEnd = Math.floor(currentRage / 2) + currentStart;
```
```suggestion
                    currentEnd = currentStart + Math.floor(currentRange / 2) ;
```

sisou commented on src/ethers.ts:
```diff
@@ -277,6 +277,63 @@ async function updateUsdtBridgedBalances(addresses: string[] = [...usdtBridgedBa
+                    console.warn('Query filter failed retrying with rage of 2000.', err, currentStart, currentEnd);
+                } else {
+                    currentEnd = Math.floor(currentRage / 2) + currentStart;
+                    console.warn('Query filter failed retrying with halved range.', err, currentStart, currentEnd);
```
```suggestion
                    console.warn('Query filter failed. Retrying with halved range.', err, currentStart, currentEnd);
```

sisou reviewed (commented):
Thanks @ii-cruz!

I'll make some adjustments and then try to get it out ASAP :+1:

sisou committed:
Some spelling corrections and formatting

sisou committed:
Halve the initial Alchemy query filter block range

Recently, all queries with the old range started to fail and recommended this smaller range instead.

sisou committed:
When query filter fails, halve the range and retry

Until the documented maximum no-log-limit range of 2000.
